### PR TITLE
Routing with bidi

### LINKED ISSUE #2: controller namespaces must be referenced in views even if not used
Brevity currently uses `defroutes` in the controller to hydrate a model. This is great.

However, these controller namespaces _must_ be referenced somewhere, otherwise the `defroute` never registers. You now have functionality that is silently tied to (and therefore breaks) based on whether a certain namespace is required.

One option is to move all defroutes to something like `controller.core` and declare them all there. 

Another option is to find a way to make sure all controllers are required automagically (that's probably not possible?).

### LINKED ISSUE #4: routing/navigation is too complicated
_I discussed this with @jmdx and @spencerreeves_

### TL;DR
SPA Routing/Navigation is a hard problem.

Many have failed before us.

If we get it right, that would be good.

### List of Issues
1. **Secretary is old** The client-side routing library we're using, is old, and mostly unmaintained: https://github.com/gf3/secretary.
2. **It's too complicated to add a route** To get a route working you have to make at least 5 changes:
- white list the route in the back end (in itself, this one is not hard to fix, but we haven't. you just whitelist all non-api routes?)
- add the route to the testing framework
- add a controller in the front-end that uses secretary to handling navigate-to-
- **WEIRD** include the controller namespace in the view, otherwise it will silently fail (!! ☹️)
- make sure you've properly handled the cases where that page is independently navigated to (is all the state you expect to be there loaded?).

### Some features I'd like to see
- don't break the back button, so `goog.History`
- even as a SPA, you should be able to link to any page
- sensible state management
- no double-do in front-end and back-end

### PR DESCRIPTION
**Pull request by jmdx**
With this approach, route definitions appear in three places:
- `cljc/routes.cljc`: The only place actual URLs live.  This file associates each URL with a clojure keyword, with separate lists for API urls and page URLs.
- `clj/routes/core.clj`: The api-views def maps keywords to API handlers.  Handlers themselves are more or less the same as they were before.
- `cljs/views/core.cljs`: This views def maps keywords to page components.  

Some benefits of this:
- On the backend, we no longer have to manually list the pages to route to index.html.
- There's a single source of truth for URLs.
- Fewer macros
- If we want to attach metadata like protobuf types to API definitions, `clj/routes.cljc` would be a really good place to do it. 

(Fixes #4, fixes #2)

### CODE REVIEW

**Comment by chairmanwow** [comment]
Man, bidi’s DSL is so ugly, only a mother could love it. Is there any way we can wrap that into a better DSL using a wrapper namespace? I love this concept though!

**Comment by chairmanwow** [comment]
Also is there any way to use namespaced keywords or something like `:blog/post` rather than `:blog-post`?

File: resources/leiningen/new/brevity/src/brevity/clj/routes/middleware.clj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+(ns {{name}}.clj.routes.middleware
+  (:require [bidi.bidi :as bidi]))
+
+(defn wrap-bidi [handler wrapped-routes renderer]
+      (fn [{:keys [uri path-info] :as req}]
+          (let [path (or path-info uri)
+                resolved-route (bidi/match-route* wrapped-routes path req)
+                resolved-handler (:handler resolved-route)
+                route-params (:route-params resolved-route)]
+               (if (= :four-o-four resolved-handler)
```
Comment (chairmanwow) [inline resources/leiningen/new/brevity/src/brevity/clj/routes/middleware.clj]: Should we just have it return 404 if there’s no `resolved-handler`?

File: resources/leiningen/new/brevity/src/brevity/cljs/views/blog.cljs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+(ns {{name}}.cljs.views.blog
+  (:require [{{name}}.cljc.routes :as routes]))
+
+(def blog-posts
+  [{:id "interesting-content" :title "A very interesting post" :content "Coming soon!"}
+   {:id "halting-problem" :title "How to solve the halting problem" :content "Not coming any time soon."}])
+
+(defn blog [params]
+      [:span
```
Comment (chairmanwow) [inline resources/leiningen/new/brevity/src/brevity/cljs/views/blog.cljs]: Prefer `div` over `span` unless you need `display:inline`

File: resources/leiningen/new/brevity/src/brevity/clj/routes/middleware.clj (hunk 1 above)
Comment (jmdx) [reply resources/leiningen/new/brevity/src/brevity/clj/routes/middleware.clj]: In that case it'll go down the chain - going to static resources and eventually the not-found handler.  I just made a change that lets the not-found handler still render the page with a 404, which will fall through to whatever component is mapped on the frontend for `:four-o-four`.

Comment (chairmanwow) [reply resources/leiningen/new/brevity/src/brevity/clj/routes/middleware.clj]: What about just throwing an exception and clojure.lang.ExceptionInfo? I've been debating whether or not we should just have 4**'s throwing exceptions. Thoughts?

File: resources/leiningen/new/brevity/src/brevity/cljc/routes.cljc [hunk 3]
```diff
@@ -0,0 +1,19 @@
+(ns {{name}}.cljc.routes
+    (:require [bidi.bidi :as bidi]
+              [{{name}}.cljc.utils :as u]))
+
+(def page-routes
+  (u/router "/"
+            (u/page "" :index)
+            (u/context "blog"
+                       (u/page "/" :blog)
+                       (u/page ["/entry/" :id] :blog/entry))))
```
Comment (chairmanwow) [inline resources/leiningen/new/brevity/src/brevity/cljc/routes.cljc]: What are your thoughts on namespaced keywords? I kinda hate how these keywords don't work super well in `let` in Clojure. For example,
```clojure
(let [{:keys [hello/world]} {:hello/world 4}] 
   hello/world)

; => CompilerException java.lang.RuntimeException: No such namespace: hello, compiling:(null:1:1)
```

Comment (chairmanwow) [reply resources/leiningen/new/brevity/src/brevity/cljc/routes.cljc]: Also, kickass job with this DSL. Much clearer.

Comment (jmdx) [reply resources/leiningen/new/brevity/src/brevity/cljc/routes.cljc]: Thanks!  I'm kind of on the fence - I can't think of too many cases where a map keyed by route keywords would be passed around, but if someone ran into such a case the compiler errors would be frustrating.  I think they would add a lot of value as a project grows, though - I like `:blog/themes/banner-image/upload` a lot better than `:blog-themes-banner-image-upload` or `:banner-image-upload`.

File: resources/leiningen/new/brevity/src/brevity/clj/routes/middleware.clj (hunk 1 above)
Comment (jmdx) [reply resources/leiningen/new/brevity/src/brevity/clj/routes/middleware.clj]: In this case if a route isn't found, the middleware is relying on the fallthrough to `compojure.route/resources`.  I like the idea of completely eliminating 500's in production though - probably in the course of #6.

File: resources/leiningen/new/brevity/src/brevity/cljc/routes.cljc (hunk 3 above)
Comment (jmdx) [reply resources/leiningen/new/brevity/src/brevity/cljc/routes.cljc]: It turns out `:blog/themes/banner-image/upload` isn't actually a valid keyword - so I'm gravitating towards just pure kebab keywords to minimize surprises for people.

**Pull request closed without merging**